drivers: watchdog: wdt_mcux_wwdt: fix window calculation and warning interrupt handling

Holt-Sun opened the pull request:
This PR fixes several issues in the NXP WWDT driver and adds board-specific configuration for mimxrt700_evk:

### Summary of Changes

1. **Fix WWDT window value calculation** (Patch 2)
   - Corrected window calculation formula: `windowValue = timeoutValue - min_window_ticks`
   - Added proper range validation for timeout values (MIN_TIMEOUT to MAX_TIMEOUT)
   - Previous implementation incorrectly set window value to minimum timeout directly

2. **Reject callback without warning interrupt** (Patch 3)
   - Return `-ENOTSUP` when callback is provided but `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG` is 0
   - Added error logging to clarify configuration requirements
   - Prevents silent no-op when callback would never fire

3. **Configure ITRC for WWDT reset handling on mimxrt700_evk** (Patch 4)
   - Reconfigure ITRC OUT3_SEL0/OUT4_SEL0 registers to 0xAAAAAA0A
   - Fixes issue where ROM configuration causes ITRC to reset system faster than WWDT
   - Allows RSTCTL to properly handle WWDT reset events and warning interrupts
   - Enables warning interrupt in board test configuration

4. **Enable write-through cache for mimxrt700 watchdog tests** (Patch 1)
   - Adds `CONFIG_ARM_MPU_SRAM_WRITE_THROUGH=y` to ensure timely data flush to SRAM

hakehuang commented:
board test pass on v4.3.0-940-gc9b01e8257db

Holt-Sun committed:
boards: mimxrt700_evk: configure ITRC for WWDT reset handling.

Fix WWDT warning interrupt by configuring ITRC to not intercept WWDT
timeout events.

ROM may configure ITRC OUT3_SEL0/OUT4_SEL0 to route WWDT0-2 to ITRC,
causing ITRC to reset the system faster than WWDT, preventing the
warning interrupt from functioning. Reconfigure these registers to
0xAAAAAA0A to allow RSTCTL to properly handle WWDT reset events.

Also enable WWDT warning interrupt in the board configuration to ensure
proper watchdog interrupt handling during tests.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Holt-Sun committed:
drivers: watchdog: wdt_mcux_wwdt: fix window calculation.

Correct the WWDT window value calculation and add proper
range validation.

The window value was incorrectly set to the minimum timeout value.
Per hardware specification, it should be calculated as:
  windowValue = timeoutValue - min_window_ticks

Also add:
- MAX_TIMEOUT boundary checks
- Validation before value assignment

Signed-off-by: Holt Sun <holt.sun@nxp.com>

hakehuang commented on tests/drivers/watchdog/wdt_basic_api/testcase.yaml:
```diff
@@ -169,3 +169,9 @@ tests:
       - CONFIG_PM=y
       - CONFIG_PM_DEVICE=y
+  drivers.watchdog.basic_api.dcache:
+    platform_allow:
```
if you add a case for this you need add a filter see https://github.com/zephyrproject-rtos/zephyr/blob/c3190d34aa912d39cd082db822f6ebb332806a2b/tests/drivers/uart/uart_async_api/testcase.yaml#L66

also you need filter out other cases for rt700.

Holt-Sun committed:
tests: watchdog: use write-through for mimxrt700 board.

Add CONFIG_ARM_MPU_SRAM_WRITE_THROUGH=y to
make sure data will be flushed into SRAM timely.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

Holt-Sun commented:
Add @zejiang0jason as assignee because he is the collaborator of NXP platform drivers and take lots of routine maintenance work.

zejiang0jason commented on drivers/watchdog/wdt_mcux_wwdt.c:
```diff
@@ -107,31 +108,37 @@ static int mcux_wwdt_install_timeout(const struct device *dev,
+			LOG_ERR("Warning interrupt callback requires "
+				"CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG > 0");
+			return -ENOTSUP;
+		}
```
suggestions:
1. Seems the callback change is not related with the WWDT window improvement, suggest use a separate patch.
2. Add some comment to explain why do like this.

Holt-Sun committed:
drivers: watchdog: wdt_mcux_wwdt: reject callback without warning IRQ

The WWDT callback is only invoked from the warning interrupt path.
If CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG is 0, the warning
interrupt is effectively disabled, so a configured callback would
never fire.
Return -ENOTSUP and log an error when a callback is provided
without enabling the warning interrupt, to avoid a silent no-op.

Signed-off-by: Holt Sun <holt.sun@nxp.com>

HyperionCoding commented:
This change (and the previous one that added `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG` requirement) made `WDT_FLAG_RESET_NONE` unusable. For debugging I used "callback only" timeout by setting flags to `WDT_FLAG_RESET_NONE` and callback to my callback that prints the current thread etc. Having to set `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG` means that every timeout is now two timeouts. I could work around it by setting the actual watchdog timeout to the warning length but the kconfig `range 0 1023` limits it lower than I want.

It is an opinionated question whether Zephyr should make the callback warning only or also the timeout but to restore the previous (and full) functionality setting callback without warning should be allowed.

Holt-Sun commented:
> This change (and the previous one that added `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG` requirement) made `WDT_FLAG_RESET_NONE` unusable. For debugging I used "callback only" timeout by setting flags to `WDT_FLAG_RESET_NONE` and callback to my callback that prints the current thread etc. Having to set `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG` means that every timeout is now two timeouts. I could work around it by setting the actual watchdog timeout to the warning length but the kconfig `range 0 1023` limits it lower than I want.
> 
> It is an opinionated question whether Zephyr should make the callback warning only or also the timeout but to restore the previous (and full) functionality setting callback without warning should be allowed.

@HyperionCoding, thanks, I think the concern on the reset-none path is valid, but it predates PR #101271. The `-ENOTSUP` restriction was introduced earlier in dbbdd3c6cfe; #101271 mainly kept that behavior and clarified the logging.

For `WDT_FLAG_RESET_NONE`, allowing a callback with `CONFIG_WDT_MCUX_WWDT_WARNING_INTERRUPT_CFG=0` restores the regular callback-at-expiry flow without introducing an extra warning stage. For reset-enabled modes, keeping the warning requirement still makes sense so the callback can run before reset.

I’ve prepared that change in https://github.com/zephyrproject-rtos/zephyr/pull/107188
